### PR TITLE
[10.x] Fix Postgres cache store failed to put exist cache in transaction

### PR DESCRIPTION
**Pull request by xdevor**
## Problem
When using Postgres as the Cache database store, 
an error occurs when performing put operation on an already existing cache key during a transaction.
The error is as follows: `SQLSTATE[25P02]: In failed sql transaction: 7 ERROR: current transaction is aborted, commands ignored until end of transaction block...`

## Steps To Reproduce
- Step 1: after setup postgres and .env file put the following code to laravel `routes/console.php`
```php
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;

Artisan::command('test', function () {

    $store = Cache::store('database');

    DB::beginTransaction();

    $store->put('test-key', 'test-value');

    DB::commit();

});
```
- Step 2:  run `php artisan test`  **twice times** and the below error will show
Screen Shot 2023-11-10 at 10 39 39 PM

## Versions
- Postgres: v12 and v16
- PHP: 8.2
- Laravel: 10

### CODE REVIEW

File: tests/Cache/CacheDatabaseStoreTest.php [hunk 1]
```diff
@@ -91,10 +91,24 @@ public function testValueIsUpdatedWhenInsertThrowsException()
...
+    {
+        $store = $this->getMockBuilder(DatabaseStore::class)->onlyMethods(['getTime'])->setConstructorArgs($this->getPostgresMocks())->getMock();
+        $table = m::mock(stdClass::class);
+        $store->getConnection()->shouldReceive('transactionLevel')->once()->andReturn(1);
+        $store->getConnection()->shouldReceive('table')->once()->with('table')->andReturn($table);
+        $store->expects($this->once())->method('getTime')->willReturn(1);
+        $table->shouldReceive('upsert')->once()->with(['key' => 'prefixfoo', 'value' => base64_encode(serialize("\0")), 'expiration' => 61], 'key')->andReturnTrue();
+
+        $result = $store->put('foo', "\0", 60);
+        $this->assertTrue($result);
```
[the remark this answers is not in the document]
Comment (xdevor) [reply tests/Cache/CacheDatabaseStoreTest.php:104]: No problem, I've written an integration test to prove the bug exist and  also the integration test validity [here](https://github.com/xdevor/framework/pull/2). If the implementation of the test case is fine, I will push the commit here.

**Comment by tpetry** [comment]
Wouldn't an upsert for all database drivers make the most sense? For existing keys they will have to execute two queries instead of just one with upserts.

**Comment by xdevor** [comment]
@tpetry I 100% agree with your point, but the reason why I didn't do that in this bug fix PR is because I'm want to minimize side effect, also i'm not entirely sure if there has some reason that led to the previous implementation not using upsert.

**Comment by tpetry** [comment]
@plehatron Was there a specific reason for not using upsert?

**Comment by xdevor** [comment]
If using the `upsert()` approach is confirmed, I would like to consider adjusting the operation of the `$store->add()` method in database cache store to also use `upsert()`, because the current implementation of `add()` has a problem that **it does not update expired cache**. However, **logically, an expired key should be treated as non-existent** cache and should be directly updated as a whole.

```php
    public function add($key, $value, $seconds)
    {
       ...
        try {
            return $this->table()->insert(compact('key', 'value', 'expiration'));
        } catch (QueryException) {
            return $this->table()
                ->where('key', $key)
                ->where('expiration', '<=', $this->getTime()) // <--- not update expired cache
                ->update([
                    'value' => $value,
                    'expiration' => $expiration,
                ]) >= 1;
        }
    }
```

File: tests/Cache/CacheDatabaseStoreTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (xdevor) [reply tests/Cache/CacheDatabaseStoreTest.php]: Okay, I have adjusted it to integration test

**Comment by xdevor** [comment]
It seems there's no issue with using `upsert` for databases supported by Laravel. I've adjusted it to use `upsert`.

**Comment by tpetry** [comment]
The `add` method has the same problem with PG when the `insert` fails because of a duplicate key. But honestly, I don't understand the code in the exception case. The logic seems to be different than described in the docblock.

**Comment by plehatron** [comment]
> @plehatron Was there a specific reason for not using upsert?

Hey @tpetry, the original logic for `DatabaseStore::put` predates my work from https://github.com/laravel/framework/pull/26726 where I made sure the cache component is fully compliant with the PSR-16 SimpleCache interface. Maybe @taylorotwell has more insight into why upsert was not used then 🤷‍♂️.

**Comment by xdevor** [comment]
> The `add` method has the same problem with PG when the `insert` fails because of a duplicate key. But honestly, I don't understand the code in the exception case. The logic seems to be different than described in the docblock.

@tpetry seems like the `add()` method is designed to ensure that the cache value is not overwritten when the cache exists. However, it is challenging to prevent race conditions without lock (and lock has risk), that's why the current implementation is as it is i think.

Perhaps the only solution for now is to delete expired caches either before each `add()` call or periodically delete expired cache.

**Comment by tpetry** [comment]
> @tpetry seems like the `add()` method is designed to ensure that the cache value is not overwritten when the cache exists. However, it is challenging to prevent race conditions without lock (and lock has risk), that's why the current implementation like this.

But the logic in the catch-part is overwriting any value when the insert fails with an exception due to an existing value?

**Comment by xdevor** [comment]
> But the logic in the catch-part is overwriting any value when the insert fails with an exception due to an existing value?

Because expired cache (`where('expiration', '<=', $this->getTime())` ) is same as not existing, and db will not remove the expired record automatically.

```php
        } catch (QueryException) {
            return $this->table()
                ->where('key', $key)
                ->where('expiration', '<=', $this->getTime()) // <--- only update expired cache, and expired cache is to be considered as not existing
                ->update([
                    'value' => $value,
                    'expiration' => $expiration,
                ]) >= 1;
        }
```

**Comment by xdevor** [comment]
> Oh you're right. didn't look could enough. Still the `current transaction is aborted` would also happen if the `insert` fails, an `insertOrIgnore` needs to be used here.

@tpetry oh seems like possible ~ maybe implement like this to make sure it will not be silent while race condition occur, and call `get()` before insert  because `get()` will delete expired cache automatically
```php
        throw_if(
            is_null($this->table()->get($key))
            && $this->table()->insertOrIgnore(compact('key', 'value', 'expiration')) === 0,
            new \RuntimeException("Cache key '$key' has been set.")
        )
        return true;
```

**Pull request merged**